owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/ConnectionValidator.kt

```diff
@@ -0,0 +1,187 @@
+
+import android.accounts.AccountManager
+import android.accounts.AccountsException
+import android.content.Context
+import com.owncloud.android.lib.common.authentication.OwnCloudCredentials
+import com.owncloud.android.lib.common.authentication.OwnCloudCredentialsFactory.OwnCloudAnonymousCredentials
+import com.owncloud.android.lib.common.http.HttpConstants
+import com.owncloud.android.lib.common.operations.RemoteOperationResult
+import com.owncloud.android.lib.resources.files.CheckPathExistenceRemoteOperation
+import com.owncloud.android.lib.resources.status.GetRemoteStatusOperation
+import com.owncloud.android.lib.resources.status.RemoteServerInfo
+import org.apache.commons.lang3.exception.ExceptionUtils
+import timber.log.Timber
+import java.io.IOException
+import java.lang.Exception
+
+class ConnectionValidator(
+    val context: Context,
+    val clearCookiesOnValidation: Boolean
+) {
+
+    fun validate(baseClient: OwnCloudClient, singleSessionManager: SingleSessionManager): Boolean {
+        try {
+            var validationRetryCount = 0
+            val client = OwnCloudClient(baseClient.baseUri, null, false, singleSessionManager)
+            if (clearCookiesOnValidation) {
+                client.clearCookies()
+            } else {
+                client.cookiesForBaseUri = baseClient.cookiesForBaseUri
+            }
+
+            client.account = baseClient.account
+            client.credentials = baseClient.credentials
+            while (validationRetryCount < VALIDATION_RETRY_COUNT) {
+                Timber.d("validationRetryCout %d", validationRetryCount)
+                var successCounter = 0
+                var failCounter = 0
+
+                client.setFollowRedirects(true)
+                if (isOnwCloudStatusOk(client)) {
+                    successCounter++
+                } else {
+                    failCounter++
+                }
+
+                // Skip the part where we try to check if we can access the parts where we have to be logged in... if we are not logged in
+                if (baseClient.credentials !is OwnCloudAnonymousCredentials) {
+                    client.setFollowRedirects(false)
+                    val contentReply = canAccessRootFolder(client)
+                    if (contentReply.httpCode == HttpConstants.HTTP_OK) {
+                        if (contentReply.data == true) { //if data is true it means that the content reply was ok
+                            successCounter++
+                        } else {
+                            failCounter++
+                        }
+                    } else {
+                        failCounter++
+                        if (contentReply.httpCode == HttpConstants.HTTP_UNAUTHORIZED) {
+                            checkUnauthorizedAccess(client, singleSessionManager, contentReply.httpCode)
+                        }
+                    }
+                }
+                if (successCounter >= failCounter) {
```

Comment:
Would it be possible to fail 2 times, and succeed on the third one? What is the expected behaviour in that case?